File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        stack = deque()
        curr = root
        
        while curr or stack:
            # add all left nodes of curr to the stack
            while curr:
                stack.append(curr)
                curr = curr.left
            
            curr = stack.pop()
            ans.append(curr.val)
            curr = curr.right
        
        return ans

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.LinkedList;
import java.util.Stack;

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        Stack<TreeNode> s = new Stack<>();
        TreeNode currNode = root;
        boolean done = false;
        
        while (!done) {
            if (currNode != null) {
                s.push(currNode);
                currNode = currNode.left;
            } else {
                if (s.isEmpty()) {
                    done = true;
                } else {
                    currNode = s.pop();
                    res.add(currNode.val);
                    currNode = currNode.right;
                }
            }
        }
        
        return res;
    }
}